### PR TITLE
Added env command and refactored config to use multi-auth

### PR DESCRIPTION
**Pull request by alijared**

### CODE REVIEW

File: croud/util.py [hunk 1]
```diff
@@ -76,3 +76,14 @@ def open_page_in_browser(url: str) -> int:
...
     return webbrowser.open_new_tab(url)
+
+
+def nested_dict_value(data: dict, key: str):
+    cursor = data
+    for k in key.split("."):
+        if k in cursor:
+            cursor = cursor[k]
+        else:
+            return None
```
Comment (MarkusH) [inline croud/util.py]: I feel like this should raise a `KeyError` or something instead of only returning `None`.

File: croud/config.py [hunk 2]
```diff
@@ -22,35 +22,59 @@
...
 
+from croud.util import nested_dict_value
+
 
 class Configuration:
     USER_CONFIG_DIR: str = user_config_dir("Crate")
     FILENAME: str = "croud.yaml"
     FILEPATH: str = f"{USER_CONFIG_DIR}/{FILENAME}"
+    DEFAULT_CONFIG: dict = {
+        "auth": {"currentContext": 0, "contexts": [{"env": "prod", "token": ""}]}
```
Comment (MarkusH) [inline croud/config.py]: Could you use snake_case over camelCase please :)
```suggestion
        "auth": {"current_context": 0, "contexts": [{"env": "prod", "token": ""}]}
```

File: tests/unit_tests/test_util.py [hunk 3]
```diff
@@ -86,3 +87,8 @@ def test_can_launch_browser(self, webbrowser_get_mock, get_platform_mock):
             env_mock.get.return_value = "foo"
             result = can_launch_browser()
             self.assertFalse(result)
+
+    def test_nested_dict_value(self):
```
Comment (MarkusH) [inline tests/unit_tests/test_util.py]: Can you add a test for the absence of a nested key as well as when the value of a key isn't a dictionary (or rather, doesn't support key lookup).

Something  like this:
```python
data = {"key": {"nestedkey": {"reallynestedkey": "thenestedvalue"}}}
key = nested_dict_value(data, "key.nestedkey.doesnotexist")
```
as well as
```python
data = {"key": {"nestedkey": 1}}
key = nested_dict_value(data, "key.nestedkey.foo")
```

File: tests/unit_tests/test_util.py [hunk 4]
```diff
@@ -86,3 +87,8 @@ def test_can_launch_browser(self, webbrowser_get_mock, get_platform_mock):
...
+        data = {"key": {"nestedkey": {"reallynestedkey": "thenestedkey"}}}
```
Comment (MarkusH) [inline tests/unit_tests/test_util.py]:
```suggestion
        data = {"key": {"nestedkey": {"reallynestedkey": "thenestedvalue"}}}
```

File: croud/config.py [hunk 5]
```diff
@@ -63,7 +87,7 @@ def write_config(config: dict) -> None:
         yaml.dump(config, f, default_flow_style=False, allow_unicode=True)
 
 
-def set_property(property: str, value: str):
```
Comment (MarkusH) [inline croud/config.py]: I believe there's still merit to keep this method. There will be other commands that make use of this, I think. At least the `login` command, I think.

File: CHANGES.rst [hunk 6]
```diff
@@ -14,3 +14,5 @@ Unreleased
 - Added a subcommand `logout` that clears the stored login token, and clears the OAuth session
 
 - The environment used to logged to in is now stored in config, so it is only necessary to provide the ``--env`` flag on the ``login`` subcommand
+
+- Added a subcommand `env` that allows you to switch env (so you can be logged into multiple environments. E.g. ``croud env prod``
```
Comment (jonas-arkulpa) [inline CHANGES.rst]: We normally add new changes at the top of the list.

File: croud/config.py [hunk 7]
```diff
@@ -22,35 +22,59 @@
...
             os.makedirs(Configuration.USER_CONFIG_DIR)
 
         if not os.path.isfile(Configuration.FILEPATH):
-            write_config({"env": "prod", "token": ""})
+            write_config(Configuration.DEFAULT_CONFIG)
 
     @staticmethod
     def get_env() -> str:
-        return load_config().get("env") or "prod"
+        return get_auth_context().get("env") or "prod"
```
Comment (MarkusH) [inline croud/config.py]: Why don't we do this?
```suggestion
        return get_auth_context().get("env", "prod")
```

Comment (jonas-arkulpa) [inline croud/config.py]: Would it make sense to also allow this config via env variable?
Do we have this functionality for the region?

File: croud/config.py [hunk 8]
```diff
@@ -22,35 +22,59 @@
...
 
     @staticmethod
-    def set_env(env: str) -> None:
-        set_property("env", env)
+    def set_context(env: str) -> None:
+        config = load_config()
+        contexts: list = nested_dict_value(config, "auth.contexts")
```
Comment (MarkusH) [inline croud/config.py]: I think `auth["contexts"]` should be a dict like structure and `current_context` point to the key of the current context:
```python
{
    "auth": {
        "current_context": "dev",  # or "prod"
        "contexts": {
            "dev": {"token": ...},
            "prod": {"token": ...},
        },
    },
}
```

File: croud/config.py (hunk 7 above)
Comment (alijared) [reply croud/config.py]: What do you mean? This returns the env for the current context. To change you can ``croud env $env``

File: tests/unit_tests/test_util.py [hunk 9]
```diff
@@ -86,3 +87,20 @@ def test_can_launch_browser(self, webbrowser_get_mock, get_platform_mock):
...
+        key = nested_dict_value(data, "key.nestedkey.reallynestedkey")
+        self.assertEqual(key, "thenestedkey")
+
+        data = {"key": {"nestedkey": {"reallynestedkey": "thenestedvalue"}}}
+        with self.assertRaises(KeyError) as error:
+            nested_dict_value(data, "key.nestedkey.doesnotexist")
+
+        self.assertIsInstance(error.exception, KeyError)
```
Comment (MarkusH) [inline tests/unit_tests/test_util.py]: No need for `assertIsInstance` checks on `error.exception`. That's already handled by the `self.assertRaises()` :slightly_smiling_face:

File: croud/util.py [hunk 10]
```diff
@@ -76,3 +76,16 @@ def open_page_in_browser(url: str) -> int:
...
             pass
     return webbrowser.open_new_tab(url)
+
+
+def nested_dict_value(data: dict, key: str):
+    cursor = data
+    for k in key.split("."):
+        if type(cursor) is not dict:
+            raise TypeError
+        if k in cursor:
```
Comment (MarkusH) [inline croud/util.py]: No need to do a check here. Just using `cursor = cursor[key]` will raise the `KeyError` for if `key` is not in `cursor`. Unless you want to provide additional information on the `KeyError`, which might be helpful.

File: croud/util.py [hunk 11]
```diff
@@ -76,3 +76,16 @@ def open_page_in_browser(url: str) -> int:
         except FileNotFoundError:  # WSL might be too old
             pass
     return webbrowser.open_new_tab(url)
+
+
+def nested_dict_value(data: dict, key: str):
+    cursor = data
+    for k in key.split("."):
+        if type(cursor) is not dict:
+            raise TypeError
```
Comment (MarkusH) [inline croud/util.py]: I think it would be valuable to provide a detailed error message what type `cursor` actually is

File: croud/util.py [hunk 12]
```diff
@@ -76,3 +76,16 @@ def open_page_in_browser(url: str) -> int:
         except FileNotFoundError:  # WSL might be too old
             pass
     return webbrowser.open_new_tab(url)
+
+
+def nested_dict_value(data: dict, key: str):
+    cursor = data
+    for k in key.split("."):
+        if type(cursor) is not dict:
```
Comment (MarkusH) [inline croud/util.py]: The more pythonic way would be `if not isinstance(cursor, dict)`

File: tests/unit_tests/test_util.py (hunk 9 above)
Comment (alijared) [reply tests/unit_tests/test_util.py]: Thanks

**Comment by jonas-arkulpa** [comment]
I think it would be enough to save the credentials for the env separately and spare the sub command.
But we discussed the additional value of this feature the way it is implemented now.

@mikethebeer It's currently already an optional argument, right?

If it helps developers I don't mind the extra command.

**Comment by mikethebeer** [comment]
> @mikethebeer It's currently already an optional argument, right?

On the `me` subcommand, this got reomoved with https://github.com/crate/croud/pull/7 . However, we use it already in `croud login --env dev` as optional argument. So that is why I tend to do it for the other commands as well like this.

**Comment by alijared** [comment]
I disagree about the subcommand for ``env`` because as you said, it's for developers only. I don't want to have to specify an ``--env`` flag every time, and write an optional argument for every command we write. As a developer, I want to switch environments, not parse it on every command - that's cumbersome and unnecessary. If there's a way to add the optional argument to every command without having to repeat yourself, then that's not so bad, but I'd still have to either login again to work in a different environment, or parse ``--env`` on every command - which is fine now as we have one command, but in the future we won't. 

In any case, the env subcommand was 3 lines of code, and can be removed in favour of the always optional flag

File: croud/env.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+# with Crate these terms will supersede the license and you may use the
+# software solely pursuant to the terms of the relevant commercial agreement.
+
+from croud.config import Configuration
+from croud.printer import print_info
+
+
+def env(env: str):
+    Configuration.set_context(env)
+    print_info(f'Environment switched to "{env}"')
```
Comment (mikethebeer) [inline croud/env.py]:
```suggestion
    print_info(f"Environment switched to {env}")
```

File: CHANGES.rst [hunk 14]
```diff
@@ -5,6 +5,12 @@ Changes for croud
 Unreleased
 ==========
 
+- Added a tabular format to the list of possible output formats in subcommands.
```
Comment (mikethebeer) [inline CHANGES.rst]: thats not a change u made in this commit. Probably a rebase issue

File: CONTRIBUTING.rst [hunk 15]
```diff
@@ -0,0 +1,10 @@
+============
```
Comment (mikethebeer) [inline CONTRIBUTING.rst]: please rebase

File: croud/env.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+#
+# However, if you have executed another commercial license agreement
+# with Crate these terms will supersede the license and you may use the
+# software solely pursuant to the terms of the relevant commercial agreement.
+
+from croud.config import Configuration
+from croud.printer import print_info
+
+
+def env(env: str):
```
Comment (mikethebeer) [inline croud/env.py]: Pydoc missing --> so I think `croud env -h` prints nothing right now

File: croud/env.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    Configuration.set_context(env)
```
Comment (mikethebeer) [inline croud/env.py]: what happens if a user sets an irregular env? I think you need to restrict to `dev` or `prod` somewhere. Probably already in the parser?

File: croud/me.py [hunk 18]
```diff
@@ -34,7 +35,8 @@
     type=str,
 )
 @argh.arg("-o", "--output-fmt", choices=["json"], default="json", type=str)
-def me(region=None, output_fmt=None) -> None:
+@argh.arg("--env", default=None, type=str)
```
Comment (mikethebeer) [inline croud/me.py]: If you go with the optional argument `env` it would make sense to let the user chose between `dev` and `prod` to avoid irregular envs

File: croud/env.py (hunk 16 above)
Comment (mikethebeer) [reply croud/env.py]: Still missing

File: croud/env.py (hunk 17 above)
Comment (mikethebeer) [reply croud/env.py]: ☝️

File: croud/session.py [hunk 19]
```diff
@@ -58,7 +58,7 @@ def __init__(
             cookies={"session": token}, connector=conn, headers=headers
         )
 
-    async def fetch(self, query: str) -> Union[List[JsonDict], JsonDict]:
+    async def fetch(self, query: str) -> JsonDict:
```
Comment (mikethebeer) [inline croud/session.py]: why this change? I think its correct the way it is since graphql could return also a list of dictionionaries

File: croud/util.py [hunk 20]
```diff
@@ -76,3 +76,11 @@ def open_page_in_browser(url: str) -> int:
         except FileNotFoundError:  # WSL might be too old
             pass
     return webbrowser.open_new_tab(url)
+
+
+def nested_dict_value(data: dict, key: str):
```
Comment (mikethebeer) [inline croud/util.py]: as @MarkusH already point out, why is this method necessary? why don't you access the values of a dict using subscript syntax e.g. `mydict["a"]["b"]`.

File: croud/config.py [hunk 21]
```diff
@@ -22,48 +22,88 @@
...
-            write_config({"env": "prod", "token": ""})
+            write_config(Configuration.DEFAULT_CONFIG)
 
     @staticmethod
     def get_env() -> str:
-        return load_config().get("env") or "prod"
+        if Configuration._using_context is not None:
+            return Configuration._using_context
+
+        return nested_dict_value(load_config(), "auth.current_context")
```
Comment (mikethebeer) [inline croud/config.py]: as questioned already. Is `nested_dict_value` needed just for readability. Why don't you access it with `cfg["auth"]["current_context"]`

File: croud/session.py (hunk 19 above)
Comment (alijared) [reply croud/session.py]: Fails tests with ``Union[List[JsonDict], JsonDict]`` return type

File: croud/config.py (hunk 21 above)
Comment (alijared) [reply croud/config.py]: Because this is cleaner, and could be used in the future for other dicts that have nested keys

File: croud/session.py (hunk 19 above)
Comment (mikethebeer) [reply croud/session.py]: but then I think the code that calls `fetch` need to adapt its types.

File: croud/config.py (hunk 21 above)
Comment (chaudum) [reply croud/config.py]: In you really have "problems" with nested data structures, you should think about using for example `addict` https://github.com/mewwts/addict

File: croud/config.py [hunk 22]
```diff
@@ -70,17 +67,17 @@ def get_token() -> str:
     def set_token(token: str) -> None:
         config = load_config()
 
-        context = Configuration._using_context
-        if context is None:
-            context = nested_dict_value(config, "auth.current_context")
+        context = Configuration.current_context
+        if context == "":
```
Comment (mikethebeer) [inline croud/config.py]: use

```python
if not context:
    ...
```

File: croud/config.py [hunk 23]
```diff
@@ -102,8 +99,8 @@ def write_config(config: dict) -> None:
 def get_auth_context() -> dict:
     config = load_config()
 
-    context = Configuration._using_context
-    if context is None:
-        context = nested_dict_value(config, "auth.current_context")
+    context = Configuration.current_context
+    if context == "":
```
Comment (mikethebeer) [inline croud/config.py]: same here:

```python
if not context:
```

File: CHANGES.rst [hunk 24]
```diff
@@ -5,6 +5,10 @@ Changes for croud
 Unreleased
 ==========
 
+- Added ``env`` flag to commands to temporarily override auth context
+
+- Added a subcommand `env` that allows you to switch env (so you can be logged into multiple environments. E.g. ``croud env prod``
```
Comment (chaudum) [inline CHANGES.rst]: double backtick around `env`
